src/lib.rs

```diff
@@ -2128,7 +2128,12 @@ pub unsafe trait FromZeros: TryFromBytes {
         // no allocation, but `Box` does require a correct dangling pointer.
         let layout = Layout::new::<Self>();
         if layout.size() == 0 {
-            return Box::new(Self::new_zeroed());
+            // SAFETY: Contructing a Box to a ZST from a dangling pointer is
+            // explicitly allowed:
+            // https://doc.rust-lang.org/std/boxed/index.html#memory-layout
+            unsafe {
+                return Box::from_raw(NonNull::dangling().as_ptr());
+            }
```

Comment:
Can you also add a comment that explains why we do this instead of `Box::new(Self::new_zeroed())`? The argument in your commit message makes sense, but I don't think it's obvious enough that we should expect future readers to figure it out from context.